packages/belos/tpetra/test/TFQMR/test_tfqmr_diag.cpp

```diff
@@ -0,0 +1,428 @@
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+//
+// This test generates diagonal matrices for TFQMR to solve.
+//
+// NOTE: No preconditioner is used in this case.
+//
+
+// Teuchos
+#include <Teuchos_Time.hpp>
+#include <Teuchos_StandardCatchMacros.hpp>
+
+// Tpetra
+#include <Tpetra_Core.hpp>
+#include <Tpetra_Map.hpp>
+#include <Tpetra_MultiVector.hpp>
+#include <Tpetra_Vector.hpp>
+#include <Tpetra_CrsMatrix.hpp>
+
+// Belos
+#include "BelosConfigDefs.hpp"
+#include "BelosLinearProblem.hpp"
+#include "BelosTpetraAdapter.hpp"
+#include "BelosTpetraOperator.hpp"
+#include "BelosTFQMRSolMgr.hpp"
+
+using namespace Belos;
+
+//************************************************************************************************
+
+template<class MV>
+class Vector_Operator
+{
+  public:
+
+    Vector_Operator(int m_in, int n_in) : m(m_in), n(n_in) {};
+
+    virtual ~Vector_Operator() {};
+
+    virtual void operator () (const MV &x, MV &y) = 0;
+
+    int size (int dim) const { return (dim == 1) ? m : n; };
+
+  protected:
+
+    int m, n;        // an (m x n) operator
+
+  private:
+
+    // Not allowing copy construction.
+    Vector_Operator( const Vector_Operator& ): m(0), n(0) {};
+    Vector_Operator* operator=( const Vector_Operator& ) { return NULL; };
+
+};
+
+//************************************************************************************************
+
+template<class ST, class MV>
+class Diagonal_Operator : public Vector_Operator<MV>
```

Comment:
Rename to `DiagonalOperator` (if used only in this test as it seems to be the case)